### ISSUE TITLE
prepare-refseqs.pl does not finish, last thing created is empty refSeqs.json file

### ISSUE THREAD

**Issue by NeillGibson** [GMOD/jbrowse#943]
Hi,

I am trying to create a basic data directory for Jbrowse/Apollo by running the `prepare-refseqs.pl` script on a fasta file.

`/tools/bioinfo/app/apollo-2.0.6/bin/prepare-refseqs.pl --fasta /data/physical_map/input.fa --out  /data/qa/Tools/apollo/2.0.6/integration_test_organism/data/`

After a minute or so this command has created all the sequence files and a empty `refSeqs.json` file.

```
/data/qa/Tools/apollo/2.0.6/integration_test_organism/data/seq/b90/2d2/be/Chr_01-0.txt
/data/qa/Tools/apollo/2.0.6/integration_test_organism/data/seq/b90/2d2/be/Chr_01-2.txt
/data/qa/Tools/apollo/2.0.6/integration_test_organism/data/seq/b90/2d2/be/Chr_01-3.txt
...........
/data/qa/Tools/apollo/2.0.6/integration_test_organism/data/seq/506/177/8b/Chr_07-955.txt
/data/qa/Tools/apollo/2.0.6/integration_test_organism/data/seq/506/177/8b/Chr_07-956.txt
/data/qa/Tools/apollo/2.0.6/integration_test_organism/data/seq/506/177/8b/Chr_07-957.txt
/data/qa/Tools/apollo/2.0.6/integration_test_organism/data/seq/refSeqs.json
```

The the command keeps hanging, the `refSeqs.json` file never gets content written, and the expected `tracks.conf` and `trackList.json` file never get created. 

The prepare-refseqs.pl command first uses c.a. 15% cpu at this time and then drops to a continues c.a. 2% cpu usage. It sort of enters a sleep state and never finishes. I  left the command running over night and the command still was not finished, and no additional output was written. 

This issue seems to be machine specific. On 1 machine I have this issue, on another machine the command works fine. I changed the output directory, and the perl environment. But this did not help. 

Could you please advice on how to fix this issue. I guess something between opening a file handle for `refSeqs.json` and actually writing its content silently fails and causes the program to go into a sleep state. 

Thank you.

**Comment by nathandunn** [comment]
I would pots to the gmod-ajax mailing list as you'll probably get more (and better) responses: http://gmod.org/wiki/JBrowse#Contacts_and_Mailing_Lists

That being said:
1) do you have gzip installed on both machines?
2) in Apollo, if you run `./install_jbrowse.sh` does it work after that?  If not, can you post the install guide?
3) can you output the full log for running the command? 
4) if you can share the FASTA somewhere that might be helpful as well

You could also add print statements to FormatSequence.pm  writeRefSeqsJSON, but I'm not sure it would necessarily help.

**Comment by nathandunn** [comment]
Just out of curiosity are you running on a mac, High Sierra? 

I just saw #1820 come through and was wondering if it might be related.

**Comment by nathandunn** [comment]
@NeillGibson   Sorry, I meant Apollo's https://github.com/GMOD/Apollo/issues/1820  .  

Might be completely unrelated, but there seems to be a very weird interaction between MacOs High Sierra and bioperl.

**Comment by NeillGibson** [comment]
No I am not using a mac. Linux and perl version for the machine on which the commands works and the one were the commands do not work. 

```
#Machine on which commands work
cat /etc/*-release
Red Hat Enterprise Linux Server release 6.8 (Santiago)
perl -v
This is perl, v5.10.1 (*) built for x86_64-linux-thread-multi-ld

#Machine on which commands DON'T work
cat /etc/*-release
Red Hat Enterprise Linux Workstation release 7.4 (Maipo)
perl -v
This is perl 5, version 16, subversion 3 (v5.16.3) built for x86_64-linux-thread-multi
```

**Comment by nathandunn** [comment]
@NeillGibson  Is there any info on the respective bioPerl libs?    Also, if you just download JBrowse directly and run the setup.sh does it work?      

The `install_jbrowse.sh` command would be at the top-level directory from wherever you downloaded Apollo.

**Comment by NeillGibson** [comment]
Thank you for the advice. Will try to follow up later this morning / today.

**Comment by NeillGibson** [comment]
Hi @nathandunn . I did not have bioperl installed on the machine where the commands do not work. 
I manually installed bioperl but this did not solve the issue. 

I downloaded Jbrowse and ran setup.sh up to formatting some test data.  
```
sh setup.sh 
Legacy scripts wig-to-json.pl and bam-to-json.pl have removed from setup. Their functionality has been superseded by add-bam-track.pl and add-bw-track.pl. If you require the old versions, run 'setup.sh legacy'.
Installing javascript dependencies ...Detected precompiled version.
done
Gathering system information ...done
Installing Perl prerequisites ... done.

Formatting Volvox example data ...^C
```

After this I have these perl modules in my home perl5 lib directory
```
echo $PERL5LIB
/home/my_user/perl5/lib/perl5:
[my_user@by0q1y] VS $ ls -lh /home/my_user/perl5/lib/perl5
total 80K
drwxr-xr-x  3 my_user users 4.0K Nov 30 20:10 App
drwxr-xr-x 42 my_user users 4.0K Nov 30 20:14 Bio
drwxr-xr-x  2 my_user users 4.0K Nov 30 20:12 Capture
drwxr-xr-x  3 my_user users 4.0K Nov 30 20:11 Class
drwxr-xr-x  3 my_user users 4.0K Nov 30 20:09 CPAN
drwxr-xr-x  3 my_user users 4.0K Nov 30 20:11 Data
drwxr-xr-x  3 my_user users 4.0K Nov 30 20:11 Devel
drwxr-xr-x  3 my_user users 4.0K Nov 30 20:11 Exception
drwxr-xr-x  2 my_user users 4.0K Nov 30 20:11 IO
drwxr-xr-x  3 my_user users 4.0K Nov 30 20:09 JSON
drwxr-xr-x  3 my_user users 4.0K Nov 30 20:11 Module
-r--r--r--  1 my_user users  967 Nov 30 03:49 ok.pm
drwxr-xr-x  3 my_user users 4.0K Nov 30 20:09 Parse
drwxr-xr-x  2 my_user users 4.0K Nov 30 20:11 Sub
drwxr-xr-x  5 my_user users 4.0K Nov 30 20:10 TAP
drwxr-xr-x  7 my_user users 4.0K Nov 30 20:12 Test
drwxr-xr-x 10 my_user users 4.0K Nov 30 20:11 Test2
-r--r--r--  1 my_user users 6.3K Nov 30 03:49 Test2.pm
drwxr-xr-x  4 my_user users 4.0K Nov 30 20:09 x86_64-linux-thread-multi
```

This still did not solve the issue. I could not find the install_jbrowse.sh script. It is not in the directory were Jbrowse/Apollo is installed now, and also not included in a fresh download of the apollo zip file. 

Both the machines, the one with the issue, and the one without, are btw different from the actual machines where JBrowse/Apollo has been installed and is now running. One is a development machine, and the other a general execution machine, which we will use to create the organisms and tracks for Apollo.

**Comment by nathandunn** [comment]
I would try to run `prepare-refseqs.pl` from you're `jbrowse/bin` directory that were created when you run `./setup.sh`, I'm curious if its any different.

**Comment by nathandunn** [comment]
Also wondering if a tool like iostat might help.    Also, how large are the FASTA files you're trying to process.  The current process writes out a ton of small files, which may be chewing up IO.   For a large FASTA it can take quite a while, especially if you are writing to a shared drive. 

Another possibility is use faidx: http://www.htslib.org/doc/faidx.html  to create an index FASTA and then use the prepare-refseq with the index_fasta option (http://genomearchitect.readthedocs.io/en/latest/Data_loading.html?highlight=fasta#prepare-refseqs-pl).   You'll have to upgrade Apollo to 2.0.8, though to use the indexed FASTA.

**Comment by NeillGibson** [comment]
I tried to run prepare-refseqs.pl from the apollo-2.0.6/bin directory where the script is located. 
I also prepended the path to `apollo-2.0.6/extlib/lib/perl5` to the PERL5LIB environment variable. Both things did not help. 

The fasta is quite small it's a 0.4gigabase genome, the seq sub-directories are written within a minute or so. If I use the other machine successfully run the `prepare-refseqs.pl` command, then on the not working machine the `flatfile-to-json.pl` script also hangs without producing output and the `remove-track.pl` also does not work. 

Before upgrading to another apollo version  I am having a local IT colleague look once more at this issue.
